linked_list/linked_list.py

```diff
@@ -1,97 +1,197 @@
     def get_at_index(self, index):
-        pass
+        if index < 0 or self.length() < index:
+            return None
+        current_index = 0
+        current_node = self.head
+
+        # traverse list till reach index or end of list
+        while current_node:
+            # can add a condition to check the current node is less than index if above does not work
+            if current_index == index:
+                break
+            current_node = current_node.next
+            current_index += 1
+
+        return current_node.value
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: linear
+    # Space Complexity: constant
     def get_last(self):
-        pass
+        if self.head is None:
+            return None
+        current_node = self.head
+        tail_node = current_node
+        prev_node = None
+
+        # traverse list till reach end.
+        while current_node:
+            prev_node = current_node
+            current_node = current_node.next
+        tail_node = prev_node
+        return tail_node.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: linear
+    # Space Complexity: constant
     def add_last(self, value):
-        pass
+        if self.head is None:
+            self.head = value
+            return
+
+        prev_node = None
+        current_node = self.head
+        new_tail_node = Node(value)
+
+        # traverse list till reach end.
+        while current_node:
+            prev_node = current_node
+            current_node = current_node.next
+        # re-assign nodes so last node we found becomes second to last pointing to new tail and new tail points to none
+        new_tail_node = prev_node
+        current_node.next = new_tail_node
+        new_tail_node.next = None
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        if self.head is None:
+            return None
+        current_node = self.head
+        max_node = current_node
+
+        while current_node:
+            if current_node.value > max_node.value:
+                max_node = current_node
+                current_node = current_node.next
+
+        return max_node.value
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: linear
+    # Space Complexity: constant
     def delete(self, value):
-        pass
+        if self.head is None:
+            return
+        current_node = self.head
+        next_node = current_node.next
+        target_value = value
+        if current_node.value == target_value:
+            current_node = current_node.next

```

Comment:
This makes `current_node` equal to what was the second node in the list, but it doesn't redirect `self.head` to point at the second node in the list
```suggestion
            self.head = current_node.next

```